datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/pki/AlternativeNames.kt

```diff
@@ -0,0 +1,144 @@
+
+    val dnsNames: List<String>? = parseStringSANs(SubjectAltNameImplicitTags.dNSName)
+    val rfc822Names: List<String>? = parseStringSANs(SubjectAltNameImplicitTags.rfc822Name)
+    val uris: List<String>? = parseStringSANs(SubjectAltNameImplicitTags.uniformResourceIdentifier)
+
+    val ipAddresses: List<ByteArray> = extensions.filter { it.tag == SubjectAltNameImplicitTags.iPAddress }.apply {
+        forEach {
+            if (it !is Asn1Primitive)
+                throw Asn1StructuralException("Invalid iPAddress Alternative Name found: ${it.toDerHexString()}")
+            else if (it.content.size != 4 && it.content.size != 16) throw Asn1StructuralException("Invalid iPAddress Alternative Name found: ${it.toDerHexString()}")
+        }
+    }.map { (it as Asn1Primitive).content }
+
+    val directoryNames: List<List<RelativeDistinguishedName>> =
+        extensions.filter { it.tag == SubjectAltNameImplicitTags.directoryName }.apply {
+            forEach {
+                if (it !is Asn1Sequence) throw Asn1StructuralException("Invalid directoryName Alternative Name found: ${it.toDerHexString()}")
+            }
+        }.map { (it as Asn1Sequence).children.map { RelativeDistinguishedName.decodeFromTlv(it as Asn1Set) } }
+
+    val otherNames: List<Asn1Sequence> =
+        extensions.filter { it.tag == SubjectAltNameImplicitTags.otherName }.apply {
+            forEach {
+                if (it !is Asn1Sequence) throw Asn1StructuralException("Invalid otherName Alternative Name found: ${it.toDerHexString()}")
+            }
+        }.map {
+            (it as Asn1Sequence).also {
+                if (it.children.size != 2) throw Asn1StructuralException("Invalid otherName Alternative Name found (!=2 children): ${it.toDerHexString()}")
+                if (it.children.last().tag != 0u.toImplicitTag()) throw Asn1StructuralException("Invalid otherName Alternative Name found (implicit tag != 0): ${it.toDerHexString()}")
+                ObjectIdentifier.parse((it.children.first() as Asn1Primitive).content) //this throws if something is off
+            }
+        }
+    val ediPartyNames: List<Asn1Sequence> =
+        extensions.filter { it.tag == SubjectAltNameImplicitTags.ediPartyName }.apply {
+            forEach {
+                if (it !is Asn1Sequence) throw Asn1StructuralException("Invalid ediPartyName Alternative Name found: ${it.toDerHexString()}")
+            }
+        }.map {
+            (it as Asn1Sequence).also {
+                if (it.children.size > 2) throw Asn1StructuralException("Invalid partyName Alternative Name found (>2 children): ${it.toDerHexString()}")
+                if (it.children.find { it.tag != 0u.toImplicitTag() && it.tag != 1u.toImplicitTag() } != null) throw Asn1StructuralException(
+                    "Invalid partyName Alternative Name found (illegal implicit tag): ${it.toDerHexString()}"
+                )
+                //TODO: strict string parsing
```

Comment:
revelant todo?

Reply:
that is a larger TODO. we are simply a bit too lenient (As we've always been)